metadata/src/retain.rs

```diff
@@ -0,0 +1,174 @@
+        type_ids.insert(constant.ty.id());
+    }
+
+    if let Some(error) = &pallet.error {
+        type_ids.insert(error.ty.id());
+    }
+}
+
+/// Update the given type using the new type ID from the portable registry.
+///
+/// # Panics
+///
+/// Panics if the [`scale_info::PortableRegistry`] did not retain all needed types.
+fn update_type(ty: &mut UntrackedSymbol<TypeId>, map_ids: &BTreeMap<u32, u32>) {
+    let old_id = ty.id();
+    let new_id = map_ids
+        .get(&old_id)
+        .unwrap_or_else(|| panic!("PortableRegistry did not retain type id {old_id}. This is a bug. Please open an issue."));
+    *ty = (*new_id).into();
+}
+
+/// Update all type IDs of the provided pallet using the new type IDs from the portable registry.
+fn update_pallet_types(pallet: &mut PalletMetadata<PortableForm>, map_ids: &BTreeMap<u32, u32>) {
+    if let Some(storage) = &mut pallet.storage {
+        for entry in &mut storage.entries {
+            match &mut entry.ty {
+                StorageEntryType::Plain(ty) => {
+                    update_type(ty, map_ids);
+                }
+                StorageEntryType::Map { key, value, .. } => {
+                    update_type(key, map_ids);
+                    update_type(value, map_ids);
+                }
+            }
+        }
+    }
+
+    if let Some(calls) = &mut pallet.calls {
+        update_type(&mut calls.ty, map_ids);
+    }
+
+    if let Some(event) = &mut pallet.event {
+        update_type(&mut event.ty, map_ids);
+    }
+
+    for constant in &mut pallet.constants {
+        update_type(&mut constant.ty, map_ids);
+    }
+
+    if let Some(error) = &mut pallet.error {
+        update_type(&mut error.ty, map_ids);
+    }
+}
+
+/// Generate a subset of the metadata that contains only the
+/// types needed to represent the provided pallets.
+///
+/// # Note
+///
+/// Used to strip metadata of unneeded information and to reduce the
+/// binary size.
+///
+/// # Panics
+///
+/// Panics if the [`scale_info::PortableRegistry`] did not retain all needed types,
+/// or the metadata does not contain the "sp_runtime::DispatchError" type.
+pub fn retain_metadata_pallets<F>(metadata: &mut RuntimeMetadataV14, mut filter: F)
+where
+    F: FnMut(&PalletMetadata<PortableForm>) -> bool,
+{
+    let mut type_ids = HashSet::new();
+
+    // Collect all types needed to represent the provided pallets.
+    metadata.pallets.retain(|pallet| {
+        if filter(pallet) {
+            collect_pallet_types(pallet, &mut type_ids);
+            true
+        } else {
+            false
+        }
+    });
+
+    // Additionally, subxt depends on the `RuntimeError` and `DispatchError`.
+    let dispatch_error_ty = metadata
+        .types
+        .types()
+        .iter()
+        .find(|ty| ty.ty().path().segments() == ["sp_runtime", "DispatchError"])
+        .expect("Metadata must contain sp_runtime::DispatchError");
+
+    type_ids.insert(dispatch_error_ty.id());
```

Comment:
So just looking at the V14 top level type:

```rust
pub struct RuntimeMetadataV14 {
	/// Type registry containing all types used in the metadata.
	pub types: PortableRegistry,
	/// Metadata of all the pallets.
	pub pallets: Vec<PalletMetadata<PortableForm>>,
	/// Metadata of the extrinsic.
	pub extrinsic: ExtrinsicMetadata<PortableForm>,
	/// The type of the `Runtime`.
	pub ty: <PortableForm as Form>::Type,
}
```

I can see that we are retaining types from the pallets, but I can't see how we are keeping `extrinsic` and `ty` types above offhand (and yet I think they are contributing to the sizes being big so I feel like I'm missing something :))?

Reply:
My bad, I didn't push the revert on the master. I've retained the types from the `extrinsic` and `runtime_ty`